src/realm/parser/query_builder.cpp

```diff
@@ -0,0 +1,941 @@
+        int32_t nanoseconds = 0;
+        if (time_inputs.size() == 7) {
+            nanoseconds = stot<int32_t>(time_inputs[6]);
+        }
+        return get_timestamp_if_valid(seconds, nanoseconds);
+    }
+    throw std::runtime_error("Unexpected timestamp format.");
+}
+
+StringData from_base64(const std::string& input, util::StringBuffer& decode_buffer)
+{
+    // expects wrapper tokens B64"..."
+    if (input.size() < 5
+        || !(input[0] == 'B' || input[0] == 'b')
+        || input[1] != '6' || input[2] != '4'
+        || input[3] != '"' || input[input.size() - 1] != '"') {
+        throw std::runtime_error("Unexpected base64 format");
+    }
+    const size_t encoded_size = input.size() - 5;
+    size_t buffer_size = util::base64_decoded_size(encoded_size);
+    decode_buffer.resize(buffer_size);
+    StringData window(input.data() + 4, encoded_size);
+    util::Optional<size_t> decoded_size = util::base64_decode(window, decode_buffer.data(), buffer_size);
+    if (!decoded_size) {
+        throw std::runtime_error("Invalid base64 value");
+    }
+    REALM_ASSERT_DEBUG_EX(*decoded_size <= encoded_size, *decoded_size, encoded_size);
+    decode_buffer.resize(*decoded_size); // truncate
+    StringData output_window(decode_buffer.data(), decode_buffer.size());
+    return output_window;
+}
+
+// check a precondition and throw an exception if it is not met
+// this should be used iff the condition being false indicates a bug in the caller
+// of the function checking its preconditions
+#define precondition(condition, message) if (!REALM_LIKELY(condition)) { throw std::logic_error(message); }
+
+using KeyPath = std::vector<std::string>;
+KeyPath key_path_from_string(const std::string &s) {
+    std::stringstream ss(s);
+    std::string item;
+    KeyPath key_path;
+    while (std::getline(ss, item, '.')) {
+        key_path.push_back(item);
+    }
+    return key_path;
+}
+
+struct PropertyExpression
+{
+    std::vector<size_t> indexes;
+    std::function<Table *()> table_getter;
+    size_t col_ndx;
+    DataType col_type;
+
+    PropertyExpression(Query &query, const std::string &key_path_string, const std::string& object_type)
+    {
+        KeyPath key_path = key_path_from_string(key_path_string);
+        TableRef cur_table = query.get_table();;
+        for (size_t index = 0; index < key_path.size(); index++) {
+            size_t cur_col_ndx = cur_table->get_column_index(key_path[index]);
+
+            precondition(cur_col_ndx != realm::not_found,
+                         util::format("No property '%1' on object of type '%2'", key_path[index], object_type));
+
+            DataType cur_col_type = cur_table->get_column_type(cur_col_ndx);
+            if (index != key_path.size() - 1) {
+                precondition(cur_col_type == type_Link || cur_col_type == type_LinkList,
+                             util::format("Property '%1' is not a link in object of type '%2'", key_path[index], object_type));
```

Comment:
Isn't this now using the incorrect type name when links are involved? If we've followed a link, we're no longer looking at the original type but the linked type, and it is that linked type whose name should appear in this error.

Reply:
It might be worth ensuring we have tests for these error cases that include validating the expected property name and object types appear in the error messages.

Reply:
I guess we need to pass down a list of synonym pairs to translate table names to object names. I'll add some tests for these error messages. 
This synonym pattern could also potentially enable us to parse named backlink columns in the future.

Reply:
It's probably easier to just strip any `class_` prefix from the table names used in error messages, rather than inventing a more complicated approach?

Reply:
Does stripping the `class_` prefix give the user-specified name for all cases? (Is there no special translation needed even for the `class_LinkTypesObject` example?)

Reply:
Yes.